Sources/FoundationEssentials/URL/URL.swift

```diff
@@ -1349,31 +1353,43 @@ public struct URL: Equatable, Sendable, Hashable {
         }
     }
 
-    private static func windowsPath(for posixPath: String) -> String {
-        let utf8 = posixPath.utf8
-        guard utf8.count >= 4 else {
-            return posixPath
+    #if os(Windows)
+    private static func windowsPath(for urlPath: String) -> String {
+        var iter = urlPath.utf8.makeIterator()
+        guard iter.next() == ._slash else {
+            return decodeFilePath(urlPath._droppingTrailingSlashes)
+        }
+        // "C:\" is standardized to "/C:/" on initialization.
+        if let driveLetter = iter.next(), driveLetter.isAlpha,
+           iter.next() == ._colon,
+           iter.next() == ._slash {
+            // Strip trailing slashes from the path, which preserves a root "/".
+            let path = String(Substring(urlPath.utf8.dropFirst(3)))._droppingTrailingSlashes
+            // Don't include a leading slash before the drive letter
+            return "\(Unicode.Scalar(driveLetter)):\(decodeFilePath(path))"
         }
-        // "C:\" is standardized to "/C:/" on initialization
-        let array = Array(utf8)
-        if array[0] == ._slash,
-           array[1].isAlpha,
-           array[2] == ._colon,
-           array[3] == ._slash {
-            return String(Substring(utf8.dropFirst()))
+        // There are many flavors of UNC paths, so use PathIsRootW to ensure
+        // we don't strip a trailing slash that represents a root.
+        let path = decodeFilePath(urlPath)
+        return path.replacing(._slash, with: ._backslash).withCString(encodedAs: UTF16.self) { pwszPath in
+            guard !PathIsRootW(pwszPath) else {
+                return path
```

Comment:
I think that we still get one path incorrect now: `/S:`. This a non-UNC, drive-relative path. We could either resolve that to an absolute path or return it as `S:`. You can also have a more complex path such as `S:utils\build.ps1` (which I believe would be encoded as `/s:utils/build.ps1`.

Reply:
Yeah, more work definitely needs to be done to support drive-relative paths in `URL`. Currently, `S:utils\build.ps1` is actually treated as a relative path, relative to the current working directory (but not necessarily the cwd of `S:`), so `.path` won't resolve it correctly.

I think we should get the cwd of `S:` during `URL(filePath:)` initialization and use that cwd path as the base URL, then strip `S:` from the relative path. If that sounds good I can open up a follow-on PR, since it'll be slightly more involved than this change to fix `C:\` and other absolute paths' behaviors.

Reply:
I think that we should do some checking if the call needs to be done, but otherwise, this sounds like a good way forward. Doing that in a follow up is totally fine.